Mathlib/Data/Matrix/ColumnRowPartitioned.lean

```diff
@@ -0,0 +1,229 @@
+Copyright (c) 2023 Mohanad ahmed. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Mohanad Ahmed
+-/
+
+import Mathlib.Data.Matrix.Basic
+import Mathlib.Data.Matrix.Block
+import Mathlib.LinearAlgebra.Matrix.NonsingularInverse
+
+/-! # Block Matrices from Rows and Columns
+
+This file provides the basic definitions of matrices composed from columns and rows.
+The concatenation of two matrices with the same row indices can be expressed as
+`A = fromColumns A₁ A₂` the concatenation of two matrices with the same column indices
+can be expressed as `B = fromRows B₁ B₂`.
+
+We then provide a few lemmas that deal with the products of these with each other and
+with block matrices
+
+## Tags
+column matrices, row matrices, column row block matrices
+-/
+
+namespace Matrix
+
+variable {R: Type _}
+variable {M M₁ M₂ N N₁ N₂: Type _}
+variable [Fintype M][Fintype M₁][Fintype M₂]
+variable [Fintype N][Fintype N₁][Fintype N₂]
+variable [DecidableEq M][DecidableEq M₁][DecidableEq M₂]
+variable [DecidableEq N][DecidableEq N₁][DecidableEq N₂]
+
+/-- Concatenate together two matrices A₁[M₁ × N] and A₂[M₂ × N] with the same columns (N) to get a
+bigger matrix indexed by [(M₁ ⊕ M₂) × N] -/
+def fromRows (A₁ : Matrix M₁ N R) (A₂ : Matrix M₂ N R) : Matrix (M₁ ⊕ M₂) N R :=
+  Matrix.of (Sum.elim A₁ A₂)
+
+/-- Concatenate together two matrices B₁[M × N₁] and B₂[M × N₂] with the same Rows (M) to get a
+bigger matrix indexed by [M × (N₁ ⊕ N₂)] -/
+def fromColumns (B₁ : Matrix M N₁ R) (B₂ : Matrix M N₂ R) : Matrix M (N₁ ⊕ N₂) R :=
+  Matrix.of fun i => Sum.elim (B₁ i) (B₂ i)
+
+/-- Given a column partitioned matrix extract the first column -/
+def toColumns₁ (A : Matrix M (N₁ ⊕ N₂) R) : Matrix M N₁ R :=  of fun i j => (A i (Sum.inl j))
+
+/-- Given a column partitioned matrix extract the second column -/
+def toColumns₂ (A : Matrix M (N₁ ⊕ N₂) R) : Matrix M N₂ R :=  of fun i j => (A i (Sum.inr j))
+
+/-- Given a row partitioned matrix extract the first row -/
+def toRows₁ (A : Matrix (M₁ ⊕ M₂) N R) : Matrix M₁ N R :=  of fun i j => (A (Sum.inl i) j)
+
+/-- Given a row partitioned matrix extract the second row -/
+def toRows₂ (A : Matrix (M₁ ⊕ M₂) N R) : Matrix M₂ N R :=  of fun i j => (A (Sum.inr i) j)
+
+lemma fromColumns_toColumns (A : Matrix M (N₁ ⊕ N₂) R) :
+    A = fromColumns A.toColumns₁ A.toColumns₂ := by
+  unfold fromColumns toColumns₁ toColumns₂
+  funext i j
+  cases' j
+  all_goals (simp only [of_apply, Sum.elim_inl, Sum.elim_inr])
+
+lemma fromColumns_ext_iff (A₁ : Matrix M N₁ R) (A₂ : Matrix M N₂ R) (B₁ : Matrix M N₁ R)
+    (B₂ : Matrix M N₂ R) : fromColumns A₁ A₂ = fromColumns B₁ B₂ ↔ A₁ = B₁ ∧ A₂ = B₂ := by
+  simp_rw [fromColumns, ← Matrix.ext_iff, of_apply, Sum.forall, Sum.elim_inl, Sum.elim_inr]
+  exact ⟨(fun h => ⟨fun i j => (h i).1 j, fun _ _ => (h _).2 _⟩),
+    (fun h => (fun _ => ⟨(h.1 _), (h.2 _)⟩))⟩
+
+lemma fromRows_toRows (A : Matrix (M₁ ⊕ M₂) N R) : A = fromRows A.toRows₁ A.toRows₂ := by
```

Comment:
I would move this to be right after `fromColumns_toColumns`, so that each column result is followed by the corresponding result on rows.